cpp/src/join/mixed_join_semi.cu

```diff
@@ -328,243 +249,36 @@ std::unique_ptr<rmm::device_uvector<size_type>> mixed_join_semi(
-  CUDF_EXPECTS(right_conditional.num_rows() == right_equality.num_rows(),
-               "The right conditional and equality tables must have the same number of rows.");
-
-  auto const right_num_rows{right_conditional.num_rows()};
-  auto const left_num_rows{left_conditional.num_rows()};
-  auto const swap_tables = (join_type == join_kind::INNER_JOIN) && (right_num_rows > left_num_rows);
-
-  // The "outer" table is the larger of the two tables. The kernels are
-  // launched with one thread per row of the outer table, which also means that
-  // it is the probe table for the hash
-  auto const outer_num_rows{swap_tables ? right_num_rows : left_num_rows};
-
-  auto matches_per_row = std::make_unique<rmm::device_uvector<size_type>>(
-    static_cast<std::size_t>(outer_num_rows), stream, mr);
-  auto matches_per_row_span = cudf::device_span<size_type>{
-    matches_per_row->begin(), static_cast<std::size_t>(outer_num_rows)};
-
-  // We can immediately filter out cases where one table is empty. In
-  // some cases, we return all the rows of the other table with a corresponding
-  // null index for the empty table; in others, we return an empty output.
-  if (right_num_rows == 0) {
-    switch (join_type) {
-      // Left, left anti, and full all return all the row indices from left
-      // with a corresponding NULL from the right.
-      case join_kind::LEFT_ANTI_JOIN: {
-        thrust::fill(matches_per_row->begin(), matches_per_row->end(), 1);
-        return {left_num_rows, std::move(matches_per_row)};
-      }
-      // Inner and left semi joins return empty output because no matches can exist.
-      case join_kind::LEFT_SEMI_JOIN: return {0, std::move(matches_per_row)};
-      default: CUDF_FAIL("Invalid join kind."); break;
-    }
-  } else if (left_num_rows == 0) {
-    switch (join_type) {
-      // Left, left anti, left semi, and inner joins all return empty sets.
-      case join_kind::LEFT_ANTI_JOIN:
-      case join_kind::LEFT_SEMI_JOIN: {
-        thrust::fill(matches_per_row->begin(), matches_per_row->end(), 0);
-        return {0, std::move(matches_per_row)};
-      }
-      default: CUDF_FAIL("Invalid join kind."); break;
-    }
-  }
-
-  // If evaluating the expression may produce null outputs we create a nullable
-  // output column and follow the null-supporting expression evaluation code
-  // path.
-  auto const has_nulls = cudf::nullate::DYNAMIC{
-    cudf::has_nulls(left_equality) || cudf::has_nulls(right_equality) ||
-    binary_predicate.may_evaluate_null(left_conditional, right_conditional, stream)};
-
-  auto const parser = ast::detail::expression_parser{
-    binary_predicate, left_conditional, right_conditional, has_nulls, stream, mr};
-  CUDF_EXPECTS(parser.output_type().id() == type_id::BOOL8,
-               "The expression must produce a boolean output.");
+  auto gather_map = std::make_unique<rmm::device_uvector<size_type>>(probe.num_rows(), stream, mr);
```

Comment:
This latter part of the algorithm could be replaced by something roughly like this (not tested, probably has rough edges):
```cpp
// Once you have computed `flagged`...

if (join_type != detail::join_kind::LEFT_SEMI_JOIN) {
    // Probably need another variable here since the type changes
    flagged = cudf::detail::unary_operator(*flagged, cudf::unary_operator::NOT, stream);
}

auto gather_map = cudf::detail::sequence(probe.num_rows(), cudf::numeric_scalar<size_type>(0), stream);
return cudf::detail::apply_boolean_mask(gather_map, *flagged, stream, mr);
```

The advantages:
- it's shorter
- it reuses primitives that are already compiled like `unary_operator`, `sequence`, and `apply_boolean_mask`, which avoids compiling more Thrust kernels

Reply:
Okay yep, I'll compare benchmarks with this implementation

Reply:
I ran into some trouble with types here and could use some help. The function currently returns a `std::unique_ptr<rmm::device_uvector<size_type>>`, but the suggested changes would return a `std::unique_ptr<table>`. I can use `::release()` on the table, and then on the column to get the `rmm::device_buffer`. From here I can get the type erased raw pointer. I am not sure how to construct a  `device_uvector` out of this apart from manually modifying pointer location, which isnt very elegant

Reply:
ah okay yep, there isnt a very good way to get around this

Reply:
No worries, this is fine. The readability got better with the stencil copy_if. Resolving.